src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -702,6 +725,111 @@ private async Task TryNotifyClientErrorAsync(IClientProxy client, string error,
         }
     }
 
+    internal Task UpdateRootComponents(
+        (RootComponentOperation, ComponentDescriptor?)[] operations,
+        ProtectedPrerenderComponentApplicationStore store,
+        IServerComponentDeserializer serverComponentDeserializer,
+        CancellationToken cancellation)
+    {
+        Log.InitializationStarted(_logger);
+
+        return Renderer.Dispatcher.InvokeAsync(async () =>
+        {
+            var shouldClearStore = false;
+            Task[]? pendingTasks = null;
+            try
+            {
+                if (Descriptors.Count > 0)
+                {
+                    // Block updating components if they were provided during StartCircuit. This keeps
+                    // the footprint for Blazor Server closer to what it was before.
+                    throw new InvalidOperationException("UpdateRootComponents is not supported when components have" +
+                        " been provided during circuit start up.");
+                }
+                if (_isFirstUpdate)
+                {
+                    _isFirstUpdate = false;
+                    if (store != null)
+                    {
+                        shouldClearStore = true;
+                        // We only do this if we have no root components. Otherwise, the state would have been
+                        // provided during the start up process
+                        var appLifetime = _scope.ServiceProvider.GetRequiredService<ComponentStatePersistenceManager>();
+                        await appLifetime.RestoreStateAsync(store);
+                    }
+
+                    // Retrieve the circuit handlers at this point.
+                    _circuitHandlers = [.. _scope.ServiceProvider.GetServices<CircuitHandler>().OrderBy(h => h.Order)];
+                    await OnCircuitOpenedAsync(cancellation);
+                    await OnConnectionUpAsync(cancellation);
+
+                    for (var i = 0; i < operations.Length; i++)
+                    {
+                        var operation = operations[i];
+                        if (operation.Item1.Type != RootComponentOperationType.Add)
+                        {
+                            throw new InvalidOperationException($"The first set of update operations must always be of type {nameof(RootComponentOperationType.Add)}");
+                        }
+                    }
+
+                    pendingTasks = new Task[operations.Length];
```

Comment:
The use of an array, rather than `List<Task>`, is pretty surprising. There will often be gaps in the array then, since some of the operations don't add tasks. Why not use `List<Task>`, at least to avoid forcing future devs to figure out if the gaps serve some purpose?

Reply:
We are on the first update, and we just validated that all operations are `Add`, so we know the exact number of tasks we need to capture and there won't be any gaps.

Reply:
Oh right, I see. It would be more obvious if the variable name was `pendingTasksForFirstRender` rather than just `pendingTasks`, so it's clearer the code below behaves how it does due to the assumptions in the code above.